### DISCUSSION TITLE
Tips on packaging the worker with extra plugins

### DISCUSSION THREAD

**Discussion by backbeatmedia** [inlife/nexrender#862] [Q&A]
Hello I have a use case where I am adding certain additional plugins to the worker, then packaging a Windows .exe file to run on a separate server. The destination server has nothing installed other than patched After Effects.

The plugins are some I have found mentioned in the Nexrender readme and some I have built myself. Here is the snippet from @nexrender/core/src/index.js

```
require('nexrender-sanity-upload');
require('nexrender-install-font-crossplatform');
require('nexrender-action-mogrt-template');
require('nexrender-action-sanity-patch');
```

Here is the snippet from the top level package.json

```
"nexrender-action-install-font-crossplatform": "^1.0.2",
"nexrender-action-mogrt-template": "^0.0.4",
"nexrender-action-sanity-patch": "^1.0.0",
"nexrender-sanity-upload": "^1.0.4",
```

When I run the .exe it fails, complaining that it cannot require a module. The install font one is the first it encounters, so it fails there. If I remove that bit of the job.json, it fails on the next on in the job which happens to be @nexrender/action-encode.

I am sure that I have missed something obvious, basic and simple which is too fundamental to even be worth documenting. However I am new to node.js, npm and GitHub so I am struggling and I hope someone can help. When I've got this working I can test some bits I plan to contribute back to the project - namely the integration plugins for the Sanity headless CMS, and some worker options useful when running on an EC2 instance.

I have run pkg-prelink on my repository, and fiddled about with the other scripts in the package.json file. I am stumped.

**Comment by inlife** [comment]
Hey @backbeatmedia 

One thing that you've mentioned that might be the cause for the problem:
> Here is the snippet from the top-level package.json

For the nexrender-core, all of the "embedded" plugins are defined on the core's package.json: https://github.com/inlife/nexrender/blob/master/packages/nexrender-core/package.json

That is the step that might be failing. Other than that, the flow is simple: run a single command and grab the binary:
```npm run pkg```

**Reply by backbeatmedia** [reply]
Thank you - which part of that package.json needs to contain the plugins:

"dependencies" or "peerDependencies"?

**Reply by inlife** [reply]
As you can see in the current package.json, the plugins are stored in peerDependencies. That is done to allow dynamic plugin pick-up system when using npm packages. However, for your specific case, I think both "dependencies" and "peerDependencies" would work since you are building a single binary anyway.

**Reply by backbeatmedia** [reply]
OK thank you - I have added them to "dependencies" for now, and when I run npm pkg I get this error:

```
Error! This experimental syntax requires enabling the parser plugin: 'classPrivateMethods' (73:8)
  /Users/admin/git/nexrender/packages/nexrender-cli/node_modules/@nexrender/core/node_modules/jest-runner/build/index.js
```

**Reply by inlife** [reply]
Error might be related to one of the plugins you are adding. I might suggest either checking the plugin settings or alternatively, choosing different nodejs version for pkg (this is done in nexrender-worker, nexrender-cli and nexrender-server package.json files)

**Reply by backbeatmedia** [reply]
:-) OK - any chance you could give me a clue what to put in there? I tried node16 - error no version exists.

**Reply by inlife** [reply]
You could give latest LTS node-18 a spin to see if that would work

**Reply by backbeatmedia** [reply]
Thank you so if I replace "node14-win-x64" in those files with "node18-win-x64" I need do nothing else, correct? Should I install that version of node on my dev machine? I realise these are child-level questions but I am teaching myself this stuff as I go.

**Reply by backbeatmedia** [reply]
Hello again I know you're not node tech support, and this is an open source project, but I've installed node 18 on my dev machine, and altered the package.json files to 'node18-win-x64' and it fails to complete the pkg script because it can't find a version matching "node18".

**Reply by inlife** [reply]
That specific issue is related to this package, so I invite you to research their documentation, maybe you will be able to find some helpful hint there: https://github.com/vercel/pkg#readme

**Reply by backbeatmedia** [reply]
Hello I think I may have fixed this issue. In case anyone else has it: it looked like it was trying to package jest into the nexrender binary and failing because of an experimental syntax used in that module. I used the purge-cache npm script to clean the folders, then did install-child and fingers crossed the issue described above has not come back and I can build .exe files again. That said, there is now another issue coming from the "download.js" file in nexrender core where it crashes out complaining that "[]" is not a function. However I am not sure whether or not that is something to do with the job file I am submitting so more investigation is required.

**Reply by backbeatmedia** [reply]
> Hello again I know you're not node tech support, and this is an open source project, but I've installed node 18 on my dev machine, and altered the package.json files to 'node18-win-x64' and it fails to complete the pkg script because it can't find a version matching "node18".

Hello just to finish this off for anyone else having the same problem - the issue here was that `pkg` was out of date. An update fixed the problem.
